scripts/hud/customizer.ts

```diff
@@ -0,0 +1,644 @@
+// I'm trying something new here and using a namespace to escape the namespace pollution nightmare currently have
+// In short, TypeScript seems to put stuff that doesn't include `import`/`export`s in the global namespace,
+// and using those is impossible since pure V8 doesn't have an import/module system.
+
+// The below `Axis` enum is breaking compiles because we have some fucking JSDOC (!) typedef in linegraph.js
+// that overlaps with this one. So, I'm using a namespace (which just gets compiled to an IIFE for scope isolation)
+// and not using a static class - that was essentially just playing the role of an IIFE anyway!
+
+// Not sure if this is what we'll stick with it, but seems good for now.
+
+// Sasha: had to change some types from Panel to GenericPanel
+
+import * as LayoutUtil from 'common/layout';
+
+enum Axis {
+	X = 0,
+	Y = 1
+}
+
+const Axes = [Axis.X, Axis.Y] as const;
+
+type Snaps = [SnapMode, SnapMode];
+
+/**
+ * The four different snapping behaviours
+ * For a horizontal panel, min and max are left and right respectively, for vertical, top and bottom
+ */
+enum SnapMode {
+	MIN = 1,
+	MID = 2,
+	MAX = 3,
+	OFF = 4
+}
+
+const ResizeDirection = {
+	TOP: 1,
+	TOP_RIGHT: 2,
+	RIGHT: 3,
+	BOTTOM_RIGHT: 4,
+	BOTTOM: 5,
+	BOTTOM_LEFT: 6,
+	LEFT: 7,
+	TOP_LEFT: 8
+};
```

Comment:
Don't think you touched this but let's move this to an enum while we're here. There's some code below that iterates over this enum, we have some util functions in `scripts/enum.ts` you can use.

TS enum reverse-mapping is very ugly so tempted to avoid them completely, but seeing as we have existing stuff for enum iteration we can stomach the enum usage.

Reply:
Converted ResizeDirection into a enum. The way ResizeDirection gets converted into DragMode seems problematic though, using a string union might have solved this in a cleaner way.